kpatch-build fails when patched module calls new vmlinux function

7770c18f93cc breaks patches which add calls from patched modules to new functions in a patched vmlinux.  That change enabled undefined symbol checking of the build log by kpatch-build, previously these checks were silently failing.  The `module-call-external.patch` integration test now fails to build as it adds a new symbol to the kernel and calls it from the nfsd module.

There's a little bit of git history behind (allowing) undefined symbols:

- 2014-08-29 - 4dee89269c83 ("add support for shadow variables")
-- Adds "undefined reference" kbuild.log check
-- Adds `LDFLAGS_vmlinux="--warn-unresolved-symbols"` and `LDFLAGS_MODULE="--warn-unresolved-symbols"`

- 2014-10-01 - ea445c03338a ("allow shadow functions to be called from modules")
-- Replaces `LDFLAGS_MODULE` with `KBUILD_MODPOST_WARN=1`
-- Adds "undefined!" kbuild.log check

- 2015-10-03 - f5de932b8dcf ("allow patched modules to call external functions")
-- Modifies the kmod core module to look for "external" symbols, which may be exported (as previously checked before this change), or provided by another .o within the patch module
-- Adds `module-call-external.patch` to the integration tests

- 2016-12-05 - 1330b4a3e5e9 ("kpatch-build: fix 'undefined reference to kpatch_shadow_*' errors")
-- Moves `--warn-unresolved-symbols` linker option to kpatch-gcc

- 2017-11-09 - 7770c18f93cc ("kpatch-build: Fix the logfile checks for 'undefined'")
-- Makes the undefined symbol checks actually work

So I think the intention was to tolerate undefined symbols to support the shadow variable API as provided by the kpatch core module, as well as calling new vmlinux functions from patched modules.

Shadow variable API references from patched modules and patched vmlinux tested as OK.  However, I think we need to be a little more clever about how we handle cases like `module-call-external.patch`.  Perhaps we should collect the list of reported undefined symbols and compare to symbols provided by the patch before considering it a build failure?

edit: added info about f5de932 ("allow patched modules to call external functions")

--- a/kpatch-build/kpatch-build
+++ b/kpatch-build/kpatch-build
@@ -713,8 +713,13 @@ export KPATCH_GCC_TEMPDIR
 CROSS_COMPILE="$TOOLSDIR/kpatch-gcc " \
 	KBUILD_MODPOST_WARN=1 \
 	make "-j$CPUS" $TARGETS 2>&1 | logger || die
-grep "undefined reference" "$LOGFILE" | grep -qv kpatch_shadow && die
-grep "undefined!" "$LOGFILE" | grep -qv kpatch_shadow && die
+
+# source.c:(.section+0xFF): undefined reference to `symbol'
+grep "undefined reference" "$LOGFILE" | sed -r "s/^.*\`(.*)'$/\1/" \
+	>${TEMPDIR}/undefined_references
+
+# WARNING: "symbol" [path/to/module.ko] undefined!
+grep "undefined!" "$LOGFILE" | cut -d\" -f2 >>${TEMPDIR}/undefined_references
 
 if [[ ! -e "$TEMPDIR/changed_objs" ]]; then
 	die "no changed objects found"
@@ -853,6 +858,27 @@ if ! "$KPATCH_MODULE"; then
 	"$TOOLSDIR"/create-klp-module $extra_flags "$TEMPDIR/patch/tmp.ko" "$TEMPDIR/patch/$MODNAME.ko" 2>&1 | logger 1
 	check_pipe_status create-klp-module
 fi
+
+readelf --wide --symbols "$TEMPDIR/patch/$MODNAME.ko" 2>/dev/null | \
+	awk '($4=="FUNC" || $4=="OBJECT") && ($5=="GLOBAL" || $5=="WEAK") && $7!="UND" {print $NF}' \
+	>${TEMPDIR}/new_symbols
+
+if "$KPATCH_MODULE"; then
+	cat >>${TEMPDIR}/new_symbols <<-EOF
+		kpatch_shadow_free
+		kpatch_shadow_alloc
+		kpatch_register
+		kpatch_shadow_get
+		kpatch_unregister
+		kpatch_root_kobj
+	EOF
+fi
+
+# Compare undefined_references and new_symbols files and print only the first
+# column containing lines unique to first file.
+UNDEFINED=$(comm -23 <(sort -u ${TEMPDIR}/undefined_references) \
+	<(sort -u ${TEMPDIR}/new_symbols) | tr '\n' ' ')
+[[ ! -z "$UNDEFINED" ]] && die "Undefined symbols: $UNDEFINED"
 
 cp -f "$TEMPDIR/patch/$MODNAME.ko" "$BASE" || die
 

